### ISSUE TITLE
Feature request: Selective SetRequestInterception

### ISSUE THREAD

**Issue by kblok** [puppeteer/puppeteer#4624]
This one is coming from Puppeteer-Sharp.

>A scenario would be:
>* the dev wants to download a certain resource (e.g., a video, a pdf, etc.)
>* Additionally, the dev only needs to intercept those specific resource types, having predefined URL patterns

>At the moment, all interceptions are at Request stage, with a hardcoded pattern of "*", which does not allow response body interceptions, nor selective interceptions.

What do you think guys about adding an optional array of patterns on [setRequestInterception](https://github.com/GoogleChrome/puppeteer/blob/master/lib/NetworkManager.js#L132)?
It would be sent to [Fetch.enable](https://github.com/GoogleChrome/puppeteer/blob/411347cd7bb03edacf0854760712d32b0d9ba68f/lib/NetworkManager.js#L147). I don't know whether it would be useful to expose `handleAuthRequests` as well.

**Comment by aslushnikov** [comment]
> What do you think guys about adding an optional array of patterns on setRequestInterception?
It would be sent to Fetch.enable. I don't know whether it would be useful to expose handleAuthRequests as well.

@kblok We intentionally don't do this, keeping things simple seems to be beneficial for the API and our clients!

Also, we are yet to see any strong use case that will require this. Do you have any?

**Comment by syscafedevelopment** [comment]
@aslushnikov, I think there are several use cases, especially related to downloading with Puppeteer, which devs are just working their way around, with other solutions, less elegant, to say the least.

Examples:
- what is a way for a dev to download a 5GB file using Puppeteer (taking into account existing site cookies, sessions with the server, special security headers, in a streamed manner)?
- what is a way for a dev to only get the Response headers before going further with an interception?
- Chrome DevTools offers that for free within the [Fetch](https://chromedevtools.github.io/devtools-protocol/tot/Fetch) domain ([Fetch.takeResponseBodyAsStream](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#method-takeResponseBodyAsStream), [Fetch.getResponseBody](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#method-getResponseBody), etc.) - how can we make use of that functionality?

**Comment by aslushnikov** [comment]
>  I think there are several use cases, especially related to downloading with Puppeteer, which devs are just working their way around, with other solutions, less elegant, to say the least.

@syscafedevelopment How are your usecases relate to selective request interception?

**Comment by syscafedevelopment** [comment]
In order to download stuff (retrieving response body, streamed or full), optimally, for large resources (like [this](https://movietrailers.apple.com/movies/sony_pictures/men-in-black-international/men-in-black-iinternational-trailer-1_h1080p.mov)):
- one needs to use the aforementioned [Fetch ](https://chromedevtools.github.io/devtools-protocol/tot/Fetch) methods: [Fetch.takeResponseBodyAsStream](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#method-takeResponseBodyAsStream), [Fetch.getResponseBody](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#method-getResponseBody)
- these methods will work only if the interception stage ([RequestStage](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#type-RequestStage)) is set to `Response` (aka `HeadersReceived`)
- in order to set the stage at which to begin intercepting requests, one needs to specify the `requestStage` parameter/field of the [RequestPattern](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#type-RequestPattern) object while enabling interception
- a [RequestPattern](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#type-RequestPattern) object can be set/specified while calling [Fetch.enable()](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#method-enable) method, as a parameter
- in fact, multiple [RequestPattern](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#type-RequestPattern) objects can be sent as an array, customizing the interception, from multiple points of view (e.g., both URL patterns and request stages)
- for the moment, Puppeteer does not allow a custom [RequestPattern](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#type-RequestPattern) parametrization, hardcoding only a value of `"*"` for it
- this will result in defaulting interceptions at `Request` stage, for all URLs
- the described use case (i.e., downloading/retrieving response data/body) will need at least an optional parameter in the [updateProtocolRequestInterception()](https://github.com/GoogleChrome/puppeteer/blob/6860d73212d2a36fd789a9244667be574f396164/lib/NetworkManager.js#L137) method for specifying the interception stage
- the specified interception stage may then be considered while constructing the `patterns` parameter that is to be sent to the [Fetch.enable()](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#method-enable) method
- using this new optional argument, the dev can indicate the interception stage (`Request` or `Response`), but,
- while at it, you might as well allow a full array of [RequestPatterns](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#type-RequestPattern), because:
-- the dev would probably benefit from having interception made only on some URLs - probably the ones he/she wants to download, possibly not being interested in others
-- also because I can imagine the implementation would be simpler and cleaner to have an optional array of [RequestPatterns](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#type-RequestPattern) as a parameter to that function, directly to be sent to [Fetch.enable()](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#method-enable), rather than constructing it based on other flag parameters (such as one allowing only the requestStage specification)

In short, instead of:

```
this._client.send('Fetch.enable', {
          handleAuthRequests: true,
          patterns: [{urlPattern: '*'}],
        }),
```

what can be done to allow downloads, is:

```
this._client.send('Fetch.enable', {
          handleAuthRequests: true,
          patterns: [{urlPattern: '*', requestStage: ^NEW_OPTIONAL_FLAG_PARAMETER^}],
        }),
```

or, my recommendation:

```
this._client.send('Fetch.enable', {
          handleAuthRequests: true,
          patterns: ^NEW_OPTIONAL_FULL_PATTERNS_ARRAY_PARAMETER^,
        }),
```

In which case, don't forget to have this new parameter assigned to a default (as today) value, if missing.

**Comment by aslushnikov** [comment]
> In order to download stuff (retrieving response body, streamed or full), optimally, for large resources (like this):

@syscafedevelopment Let's elaborate on this for a bit. First, you can already get response body if you wish with [`response.buffer()`](https://pptr.dev/#?product=Puppeteer&version=v1.18.1&show=api-responsebuffer). 

Second, in your protocol walk-through, you discuss how we can intercept at response stage. Indeed, we can, but this discussion is orthogonal to selective interception we discuss here. There's a separate feature request for response interception - #1191.

**Comment by syscafedevelopment** [comment]
> > In order to download stuff (retrieving response body, streamed or full), optimally, for large resources (like this):
> 
> @syscafedevelopment Let's elaborate on this for a bit. First, you can already get response body if you wish with [`response.buffer()`](https://pptr.dev/#?product=Puppeteer&version=v1.18.1&show=api-responsebuffer).
> 
> Second, in your protocol walk-through, you discuss how we can intercept at response stage. Indeed, we can, but this discussion is orthogonal to selective interception we discuss here. There's a separate feature request for response interception - #1191.

The title may be misleading. The final objective would be to allow intercepting responses, indeed. That is done by allowing selective interception, since it has to do with UrlPatterns being specified while Fetch.enable(). You may call it a confusing design decision at the CDP level, but that's another story. In any case, it's not that orthogonal to #1191 . and it's actually related also to all the "allow downloads" discussions (e.g., the long awaited #299).

Regarding response.buffer() - imagine using that with multi-GB resources. In the best case scenario, the response will be loaded/parsed by the browser, in order to be displayed, then you will be able to get its buffer, iterating the bytes again (after the browser already iterated them), loading them to memory again, in one-go. After that you'll probably need to iterate them yet again to save them to disk or whatever processing you need to do. That's gonna be (at least) a RAM killer. In the worst case scenario, it will simply not work, throwing "Protocol errors", for some cases.

What if the dev never wanted the (full) response data to be transmitted back from the server? What if based only on the headers of the response, or only on some partial response data, a dev would choose to skip that resource? At response.buffer() stage, it is too late - the response data has already been fully transmitted. You can now only re-iterate it again (as given by Chrome, not by the server at that time), if you wish.

What we need is a streamed way to do it, without being first handed to the browser (the response data, that is), and that has to do with intercepting at response stage.

After the stage is being set to Response (aka HeadersReceived), the dev can make use of the [IO.read()](https://chromedevtools.github.io/devtools-protocol/tot/IO#method-read) method, in order to get the bytes of the response, before being treated by the browser, in order to save them to the disk or process them in any way, and choosing whether or not to forward them to the browser.

Moreover, when downloading comes in mind, i.e., when setting Response stage within the UrlPatterns parameter of Fetch.enable(), as a probable direct next-step for a dev, would be to also filter the URLs to only the ones needed to be downloaded (e.g., *.mov), since it's done within the same parameter object. (`patterns`). Hence the globally-encompassing title of "Selective interception". But I agree it's more than just that.
